juju/juju
# Destroy model: watch better suited collection for changes.

## Description of change

When destroying models, we used to watch only applications and machines collections. However, when we were determining if model is ready to be destroyed, we'd check collection that stored model entities references.

This introduced 2 problems:
1. If we were to identify new entities that model should worry about before self-destructing, we would miss them as we were not watching their collections.
2. There was a potential for a race - which was observed fairly reliably - between when a specific entity collection would change and when the reference for that collection would be dealt with.  

This PR changes undertaker behavior to watch for changes on collection that references model entities rather than said entities collections.

We are also now surfacing errors that we've encountered during the attempts to destroy model as well as log them.
Drive-by fixes:

## QA steps

There are several scenarios where destroy-model would chase its own tail.
The most commonly encountered one that I have used:
1. bootstrap;
2. add model;
3. deploy an application with one unit - wait for dust to settle;
4. remove that one unit - wait for dust to settle;
[you would have an empty deployed app]
5. destroy this model.
Model gets destroyed successfully.

Previously, model destruction would get into infinite loop and only eventual kill-controller will get rid of it.
Note that if you are running 'watch juju status' during model destruction, you may not observe failure to destroy.

## Documentation changes

n/a - internal change

## Bug reference

https://bugs.launchpad.net/juju/+bug/1672549
https://bugs.launchpad.net/juju/+bug/1661930
